### PR TITLE
refactor/fix: Perform autofix application in OCaml

### PR DESCRIPTION
**Pull request by nmote**
This introduces an RPC mechanism by which we can call OCaml code in semgrep-core from Python. This will allow us greater flexibility when incrementally migrating functionality to OCaml as part of the osemgrep project.

As a motivating example, I have moved autofix application to OCaml. Brandon attempted this in #9229, but was forced to leave some autofix functionality in Python. After running the initial scan in `semgrep-core`, we run a baseline scan in certain cases. We can't do that if the first call to `semgrep-core` has already applied autofixes, dirtying the repository. Therefore, we need to be able to do it outside of the main call to `semgrep-core`.

I've prepared for this change by adding additional autofix e2e tests in #9948, #9914, and #9857.

This is labeled a "refactor/fix" instead of just a refactor because this actually fixes an issue with an autofix application on a line that contains a multi-byte character (in UTF-8 encoding), where the autofix occurs after the multi-byte character. See the changed e2e test snapshots.

For more details see the README.txt file.

Test plan: Automated tests, plus manual testing of many (but not all) of the error conditions I handle related to RPC calls.

### CODE REVIEW

File: src/rpc/README.txt
Comment (IagoAbal) [inline src/rpc/README.txt:13]: Don't we have a ATD spec just for semgrep-core ? I thought this ATD was mainly for the output of Semgrep, and this is internal stuff?

**Comment by IagoAbal** [comment]
For autofix I find it just a little bit risky, although perhaps it's quite a small risk since nowadays Semgrep is probably not used much for heavy code transformation. Also according to the PR there was some autofixing done in OCaml already so I don't know in which exact cases the RPC will be performed. If you have a good sense of what uses of Semgrep can be perf-wise negatively affected by using an RPC here, you could clarify in the PR description.

File: src/rpc/README.txt
Comment (nmote) [reply src/rpc/README.txt:13]: I believe that this file is used for both. We have a lot of internal-only stuff in that file already that has to do with the interface between pysemgrep and semgrep-core.

Comment (nmote) [reply src/rpc/README.txt:13]: I initially made a separate file for this RPC but then realized that it would be useful to have access to the types defined in the main file.

**Comment by nmote** [comment]
> For autofix I find it just a little bit risky, although perhaps it's quite a small risk since nowadays Semgrep is probably not used much for heavy code transformation.

Could you clarify what the risks are that you see? To me, there are:
* Performance risks, which are minimal for autofix. The serialization costs are dwarfed by other parts of Semgrep. First, most files typically don't have findings. Second, not all findings have autofixes. Third, we already serialize all of this data and more from semgrep-core to Python in the exact same manner, and we have never viewed that as a show-stopper. Last, we only call this once per Semgrep run.
* Correctness risks, which I believe I have appropriately mitigated by thorough testing.

> If you have a good sense of what uses of Semgrep can be perf-wise negatively affected by using an RPC here, you could clarify in the PR description.

I wrote a bit more about that in the README, which I think is a better place for that information than the PR description. However, to a certain extent it requires some critical thinking on the part of the person adding a new RPC call, and I'm not sure that I can write up a set of general rules. Let me know if you have specific things that you'd like me to add to that README, though.

File: cli/src/semgrep/autofix.py
Comment (aryx) [inline cli/src/semgrep/autofix.py]: this should be out.Edit, not ocaml.Edit

File: cli/src/semgrep/autofix.py
Comment (aryx) [inline cli/src/semgrep/autofix.py]: This should be out.Edit(...)

File: src/rpc/README.txt
Comment (aryx) [reply src/rpc/README.txt:13]: yes right now atd does not have a module system (one of the TODO of martin), so the spec of the CLI output and core output is in the same file, because they share types.

File: src/rpc/README.txt
Comment (aryx) [inline src/rpc/README.txt]: not sure I agree with this one.

File: src/rpc/README.txt
Comment (aryx) [inline src/rpc/README.txt]: typo, PRC -> RPC

File: src/rpc/README.txt
Comment (aryx) [inline src/rpc/README.txt:42]: does it mean if the packet contains newlines we have a problem?

File: src/rpc/README.txt
Comment (aryx) [inline src/rpc/README.txt:53]: which is why we want capabilities for stdout!!! :) only code with the stdout capability can print on stdout.

File: src/rpc/RPC.mli
Comment (aryx) [inline src/rpc/RPC.mli:2]: you can actually pass all_caps here, but then restrict it for the callback, so for example don't pass the stdin/stdout capability to the callback so we're sure they can't interfere.

File: cli/src/semgrep/ocaml.py
Comment (aryx) [inline cli/src/semgrep/ocaml.py]: maybe better to define an actual record type for this in the .atd.

**Review by aryx — APPROVED** [review]
LGTM.
I would prefer to move the protocol part in a different file than the callback (both for the python part and the OCaml part), and maybe use more ATD records instead
of Tuples that requires to wrap/unwrap more.

**Comment by aryx** [comment]
but let's wait for the release of 1.66 to merge this.

**Comment by aryx** [comment]
I mean I'd rather have this merged after we release 1.66.0

File: src/rpc/README.txt
Comment (nmote) [reply src/rpc/README.txt]: This is the root of your comments above about preferring `out.$X` over `ocaml.$X`. I defined more ergonomic types in `ocaml.py` for clients to use, but it is a bit of duplication. The problem with the ATD generated types is they are kind of awkward to use in Pytyhon and you have to grab the `value` property. But it's not a big deal, and I'm happy to remove this and refactor accordingly.

File: src/rpc/README.txt
Comment (nmote) [reply src/rpc/README.txt]: Good catch, thanks

File: src/rpc/README.txt
Comment (nmote) [reply src/rpc/README.txt:42]: Nope! I explicitly designed it to be robust in that case. Once we know the size from reading the header, we read `size` bytes from the stream without caring what their contents are. They can have newlines.

File: src/rpc/README.txt
Comment (nmote) [reply src/rpc/README.txt:53]: Definitely.

File: src/rpc/RPC.mli
Comment (nmote) [reply src/rpc/RPC.mli:2]: Yeah, right now we don't actually plumb any caps down (but we should, since we're writing to the filesystem!), but we will definitely want to restrict them once we do.

**Comment by nmote** [comment]
Sure, I can wait.

**Comment by nmote** [comment]
> I would prefer to move the protocol part in a different file than the callback (both for the python part and the OCaml part)

I'm happy to make the changes to use the ATD types a bit more before merging. If you're okay with it, though, I'd prefer to explore different ways to organize the code in future work after merging this PR.

File: src/rpc/README.txt
Comment (aryx) [reply src/rpc/README.txt]: I agree they are a bit awkward, but there are many places in pysemgrep where we use them already, so I don't think it's a big problem if we use them also here.

File: cli/src/semgrep/autofix.py
Comment (aryx) [inline cli/src/semgrep/autofix.py:29]: would be good here to copy some parts of your PR description here, like the rational for why
we need RPC for this, that is "brandon tried in #XXX, but then the baseline scan makes it complicated ..."

File: cli/src/semgrep/ocaml.py
Comment (aryx) [inline cli/src/semgrep/ocaml.py:104]: why the intermediate variable ? you save no space here, you replace an _ by a .

File: cli/src/semgrep/ocaml.py
Comment (aryx) [inline cli/src/semgrep/ocaml.py:99]: actually we might want the RPC callback in OCaml to print on stdout sometimes ... for example if we want
to implement some of the formatter in OCaml.

We might want to share stdout and use another file descriptor for the Python<->OCaml comm.
An alternative is have the RPC callback return a message like RetStdoutString that pysemgrep would just print.

File: src/core_cli/Core_CLI.ml
Comment (aryx) [inline src/core_cli/Core_CLI.ml:618]: Not a fan of those Qualified calls ... partly because semgrep/OCaml does not handle them ...
if we look for FooBar.foo ..., then it will not find  FooBar.(foo ...)

File: cli/src/semgrep/ocaml.py
Comment (nmote) [reply cli/src/semgrep/ocaml.py:104]: Otherwise, mypy isn't convinced that the `stdin` property remains non-None after the check.

File: cli/src/semgrep/ocaml.py
Comment (nmote) [reply cli/src/semgrep/ocaml.py:99]: Yep, definitely room for future work here.

File: src/core_cli/Core_CLI.ml
Comment (nmote) [reply src/core_cli/Core_CLI.ml:618]: This is a pretty common pattern. If Semgrep can handle `open FooBar;; foo` then we should be able to extend it to handle this.

**Comment by brandonspark** [comment]
this is super cool/clean!

File: src/rpc/RPC.ml
Comment (emjin) [inline src/rpc/RPC.ml:71]: For my own understanding: which part of this blocks?

Comment (nmote) [reply src/rpc/RPC.ml:71]: The reads from stdin

**Pull request merged**
